### PR TITLE
PACA-790: Improve support for joins in DSL

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[PACA-790](https://jira.cms.gov/browse/PACA-790)

**User Story or Bug Summary:**

Expand on the current implementation of using JPA entities as primary keys in entities generated from the DSL plugins.

Many existing RIF entities use JPA objects with ManyToOne and JoinColumn annotations as part of their primary keys.  The DSL plugins have support for this feature but it has not been used yet and will need to be tested and expanded upon before the RIF mappings can be developed.

---

### What Does This PR Do?

- Expands the expressiveness of joins in the DSL by adding `class` property to specify common join use cases.
- Adds join sub-classes for:
  - array: RDA style OneToMany with set of child entities and using `mappedBy`.
  - parent: RIF entity style OneToMany with list of child entities and using `mappedBy`.
  - child: RIF entity style ManyToOne containing parent object in child entity and using `joinColumnName`.
- Adds support for `dbOnly` column mappings to simplify using `joinColumnName` in joins safely.
- Eliminates ambiguity around when to use a column's mapping name vs db name in references from joins.
- Replaces the special `array` mapping with use of the new `array` class of joins and a custom transformer named `Array`.  Makes all joins work the same way w.r.t. transformations and simplifies model.
- Fixes the SQL generator not defining `joinColumnName` columns.
- Removes some obsolete methods and parameters.

Lots of the changes are simple things like adding arguments to methods and renaming methods to fit the model changes.  Also deleted `ArrayBean` since it is obsolete now. 

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateSqlFromDslMojo.java [hunk 1]
```diff
@@ -196,10 +195,6 @@ private List<ColumnBean> getPrimaryKeyColumns(RootBean root, MappingBean mapping
     // this map ensures no column is included twice
     final var columns = new LinkedHashMap<String, ColumnBean>();
     final var table = mapping.getTable();
-    for (JoinBean join : table.getPrimaryKeyJoinBeans()) {
-      final var joinColumn = mapping.mapJoinToParentColumnBean(root, join);
-      joinColumn.ifPresent(c -> columns.put(c.getColumnName(), c));
-    }
     for (ColumnBean column : mapping.getTable().getPrimaryKeyColumnBeans()) {
       columns.put(column.getColumnName(), column);
     }
```
Comment (KyRobbins) [inline apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateSqlFromDslMojo.java:200]: I believe `root` and `table` are now no longer used in this method

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateSqlFromDslMojo.java [hunk 2]
```diff
@@ -218,10 +213,6 @@ private List<ColumnBean> getPrimaryKeyColumns(RootBean root, MappingBean mapping
   private List<ColumnBean> getAllColumns(RootBean root, MappingBean mapping) {
     // this map ensures no column is included twice
     final var columns = new LinkedHashMap<String, ColumnBean>();
-    for (JoinBean join : mapping.getNonArrayJoins()) {
-      final var joinColumn = mapping.mapJoinToParentColumnBean(root, join);
-      joinColumn.ifPresent(c -> columns.put(c.getColumnName(), c));
-    }
     for (ColumnBean column : mapping.getTable().getColumns()) {
       columns.put(column.getColumnName(), column);
     }
```
Comment (KyRobbins) [inline apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateSqlFromDslMojo.java:218]: `root` doesn't seem to be used in this method anymore

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/java/gov/cms/model/dsl/codegen/plugin/model/ModelUtilTest.java [hunk 3]
```diff
@@ -19,6 +21,10 @@ public class ModelUtilTest {
   /** Test methods that parse class and package names from strings. */
   @Test
   public void testClassNameParsing() {
+    assertFalse(ModelUtil.isValidFullClassName(null));
+    assertFalse(ModelUtil.isValidFullClassName(""));
+    assertFalse(ModelUtil.isValidFullClassName("String"));
+    assertTrue(ModelUtil.isValidFullClassName("java.lang.String"));
     assertEquals("java.lang", ModelUtil.packageName("java.lang.String"));
     assertEquals("String", ModelUtil.className("java.lang.String"));
     assertEquals(ClassName.get(String.class), ModelUtil.classType("java.lang.String"));
```
Comment (KyRobbins) [inline apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/java/gov/cms/model/dsl/codegen/plugin/model/ModelUtilTest.java:30]: These seem like they should be separate tests, being that it's testing several different methods

**Review by KyRobbins — DISMISSED** [review]

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/ValidationUtilIT/expected-errors.txt [hunk 4]
```diff
@@ -1,24 +1,27 @@
...
+    property='table.joins[0].entityClassOrMappingDefined', desc='join null to null/null', value='false' error='must be true'
+    property='table.joins[0].fieldName', desc='join null to null/null', value='null' error='must not be null'
+    property='table.joins[1].entityClassOrMappingDefined', desc='parent join null to null/null', value='false' error='must be true'
+    property='table.joins[1].entityClassOrMappingDefined', desc='parent join null to null/null', value='false' error='must be true'
+    property='table.joins[1].entityMapping', desc='parent join null to null/null', value='null' error='must not be null'
+    property='table.joins[1].fieldName', desc='parent join null to null/null', value='null' error='must not be null'
+    property='table.joins[1].mappedBy', desc='parent join null to null/null', value='null' error='must not be null'
+    property='table.joins[2].entityClassOrMappingDefined', desc='child join null to null/null', value='false' error='must be true'
+    property='table.joins[2].entityClassOrMappingDefined', desc='child join null to null/null', value='false' error='must be true'
```
Comment (daryll) [inline apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/ValidationUtilIT/expected-errors.txt:15]: seems we have some duplicate lines here.
is this on purpose ?

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/java/gov/cms/model/dsl/codegen/plugin/model/ModelUtilTest.java (hunk 3 above)
Comment (brianburton) [reply apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/java/gov/cms/model/dsl/codegen/plugin/model/ModelUtilTest.java:30]: The different methods are simple variations doing the same thing (parsing class names) so breaking them into separate methods would be excessive.

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateSqlFromDslMojo.java (hunk 2 above)
Comment (brianburton) [reply apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateSqlFromDslMojo.java]: Thanks.  I have fixed these and removed some other unused methods.

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateSqlFromDslMojo.java (hunk 1 above)
Comment (brianburton) [reply apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/main/java/gov/cms/model/dsl/codegen/plugin/GenerateSqlFromDslMojo.java]: Thanks. I have fixed these and removed some other unused methods.

**Review by daryll — APPROVED** [review]
lgtm

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateEntitiesFromDslMojoIT/expected/FissClaim.java [hunk 5]
```diff
@@ -231,13 +228,13 @@ public class FissClaim {
...
   @Builder.Default
-  private List<FissPayer> payers = new LinkedList<>();
+  private Set<FissPayer> payers = new HashSet<>();
 
   @ManyToOne(
       fetch = FetchType.EAGER
   )
   @JoinColumn(
-      name = "`mbiId`"
+      name = "`mbi_id`"
```
Comment (mackec1) [inline apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateEntitiesFromDslMojoIT/expected/FissClaim.java:237]: do we need the back-ticks now that the db columns are snake_case?

**Review by mackec1 — APPROVED** [review]

File: apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateEntitiesFromDslMojoIT/expected/FissClaim.java (hunk 5 above)
Comment (brianburton) [reply apps/bfd-model/bfd-model-dsl-codegen/bfd-model-dsl-codegen-plugin/src/test/resources/GenerateEntitiesFromDslMojoIT/expected/FissClaim.java:237]: This was a test case verifying that the feature works.  Might be worthwhile to remove the feature in a later PR.

**Pull request merged**